### PR TITLE
[GH-7260] Add check for verified email when required by server settings

### PR DESCRIPTION
**Pull request by moonmeister**
#### Summary
Checks if users email is verified before sending email if the server requires verification of email notifications.

#### Ticket Link
#7260

### CODE REVIEW

**Comment by moonmeister** [comment]
@esethna See my notes over on the issue #7260, It'll explain the issue with Scenario A. 

As for B, I had to setup my test server to relay through google to receive emails. I'd try inviting a user via email (the / option wouldn't work for me, I had to go through the menu). If you can see other emails being sent, but not this one then maybe we have an issue. I was receiving all emails correctly in my testing.

**Comment by esethna** [comment]
Thanks @moonmeister, can you help me understand in this PR what is the behavior which is being changed?

**Comment by moonmeister** [comment]
**Scenario 1:** If the server is "closed" and thus is invite only: the email verification is automatic/implicit. The user receiving the invite is never explicitly asked to verify. They are implicitly verifying by accepting the invite. Thus they don't ever have an "unverified" email.

**Scenario 2:** If the server is "open" and thus anyone can create an account the verification is explicit. Upon account creation the user is required to verify email. There is no way past this. You have zero access till the email is verified. In fact the logs show an error saying the login failed because the email is not verified. With the account unable to login and not verified I tried mentioning it in all the ways above. Despite not even showing on the autocomplete list, you can still mention the user. But under no circumstances did if ever even execute the Send Notification function for this user. I could not get it to send a new unverified user an email. This is because the user, until they are verified, is not added to a team. There must be a check somewhere that doesn't send notifications if a user is not part of a team. 

**Scenario 3:** When a user changes their email it has to be verified, but they are still have access to the web app and they can still be mentioned at their new email address pre-verification. If they logout, then upon logon they meet the same wall given to new users on an open server and are required to verify before being given access. But like I said, they can still be mentioned because they are apart of a team now. 

**Result:** Scenario 3 then becomes the only situation in which an unverified email can receive a notification. Thu sI added code in the Send Notification to check for the unverified email address. The behavior this is changing is the behavior of only Scenario 3. Thus Existing users, who have requested an address change will not receive notifications until they verify their address. Scenario 3 is also the only situation under which this error is logged.

If we want Scenario 2 to function the same as 3 then we need to figure out why people aren't sent notifications when not part of a team and/or add people to a team when their account is created as opposed to when they verify an email. 

hopefully that makes sense.

**Comment by esethna** [comment]
Thanks @moonmeister, I think the scenario we are missing above is when the user creates an account from a [team invite link](https://docs.mattermost.com/help/getting-started/managing-members.html#team-invite-link) (on an open or closed team with email verification enabled). By creating an account this way, the user is added to the team before they have verified their email address. I was able to confirm on `master` that they start receiving email notifications before the account is verified. Is this case also covered by your fix to Scenario 3?

image

**Comment by moonmeister** [comment]
@esethna It should be as I'm checking within the actual notification function. I can test this scenario though if you'd like me too.

**Comment by esethna** [comment]
Thanks @moonmeister, we should verify that case as well, feel free to use this test server running your code:

----

Spinmint test server created at: http://i-03240600e4f4ec308.spinmint.com:8065/pr7275

**System Admin Account**: Email: test@test.com | Password: passwd

Test Account 2: Email: test2@test.com | Password: passwd

Instance ID: i-03240600e4f4ec308

**Comment by moonmeister** [comment]
@esethna I can't use the server cause verification is required and I can't verify test@test.com

**Comment by esethna** [comment]
@moonmeister I've disabled email verification on the server so you should be able to get in now.

**Comment by moonmeister** [comment]
@esethna I've confirmed it correctly throws an error in those situations. 
screen shot 2017-08-29 at 10 17 03 pm

File: app/notification.go [hunk 1]
```diff
@@ -164,6 +164,16 @@ func SendNotifications(post *model.Post, team *model.Team, channel *model.Channe
 				}
 			}
 
+			//If email verification is required and user email is not verified don't send email.
+			emailVerified := true
+			l4g.Debug("Server requires email verification: %v", utils.Cfg.EmailSettings.RequireEmailVerification)
```
Comment (hmhealey) [inline app/notification.go]: You'll want to remove these debug statements

File: app/notification.go [hunk 2]
```diff
@@ -164,6 +164,16 @@ func SendNotifications(post *model.Post, team *model.Team, channel *model.Channe
...
+			l4g.Debug("User email verification status: %v", profileMap[id].EmailVerified)
+
+			if utils.Cfg.EmailSettings.RequireEmailVerification && !profileMap[id].EmailVerified {
+				l4g.Error("Skipped sending notification email to %v, address not verified. [details: user_id=%v]", profileMap[id].Email, id)
+				emailVerified = false
```
Comment (hmhealey) [inline app/notification.go]: I think you can just use a continue statement here to save us having to get the user's status below

File: app/notification.go (hunk 1 above)
Comment (moonmeister) [reply app/notification.go]: Complete

File: app/notification.go (hunk 2 above)
Comment (moonmeister) [reply app/notification.go]: Complete

**Pull request merged**